### ISSUE TITLE
IoT Hub Client with MQTT transport does not gracefully close the tlsio connection

### ISSUE THREAD

**Issue by Arty29** [Azure/azure-iot-sdk-c#607]
- **OS and version used:** FreeRTOS v9.0.0

- **SDK version used:** azure-iot-sdk-c, 1.2.8

# Description of the issue:
I am implementing the IoT Hub Client using MQTT and with a custom tlsio adapter (direct pattern).  I have discovered that it is not possible to gracefully close the connection to the IoT Hub. 

Calling `IoTHubClient_LL_Destroy()` does result in an MQTT DISCONNECT packet being queued up in my tlsio adapter for sending, however there are no subsequent `tlsio_dowork` calls made to allow the packet to be sent.  Furthermore, no attempt is made to gracefully close the socket connection.  Instead, immediately after queuing up the MQTT DISCONNECT packet, the tlsio adapter `tlsio_destroy` function is called which logs an error because the tlsio connection is being destroyed while it is in the TLSIO_STATE_EXT_OPEN state.

I took a closer look at the underlying layers and the problem starts in `DisconnectFromClient()` which is declared in _iothubtransport_mqtt_common.c_.

In `DisconnectFromClient()` a call is made to `mqtt_client_disconnect()` which creates an MQTT DISCONNECT packet and calls `sendPacketItem()` which calls  `xio_send()` which results in the packet being queued up in my tlsio adapter, and with its associated `ON_SEND_COMPLETE` callback set to `sendComplete()` (declared in _mqtt_client.c_).  With the packet now queued, control returns to `DisconnectFromClient()` which immediately calls `xio_destroy().`  The effect is to force the tlsio connection closed, which dequeues the just-queued MQTT DISCONNECT message and uses its associated callback to specify `IO_SEND_CANCELLED` as the result.

A closer look at the handler `sendComplete()` shows that if the message containing the MQTT DISCONNECT packet (or for that matter, any  message in the queue that was sent when the MQTT client `packetState` is set to `DISCONNECT_TYPE)` were to be sent successfully, the MQTT client would actually try and gracefully close the tlsio connection by calling `close_connection().`  Unfortunately, the mechanism for supporting asynchronous close operations is somewhat crude as it only provides for at most two “do_work” calls (with a 2ms sleep in between them) to allow the connection to be closed.  However, this is really a moot point because it is currently impossible for this callback to be reached in the course of sending the MQTT DISCONNECT packet.

**Comment by jebrando** [comment]
@Arty29 Yes you are correct if the user calls destroy the DISCONNECT packet is not set to the server.  This does not violate the MQTT protocol and the IoTHub server does not require it's clients to send this final packet, never the less your point stands that its probably good business to clean up connections.

The plan here will on destroy the mqtt transport will send the final message and close the socket connection.  The timing on when I will be finish with this shall depend on many factors and I'll have to make sure I have time with other priorities.  Stay tuned and I'll update this I work on it.

**Comment by Arty29** [comment]
@jebrando Thank you for the quick response and your attention to this matter.  I will definitely stay tuned to see the results of your efforts.

**Comment by Arty29** [comment]
Hi @jebrando . Thank you for putting time into this.  Unfortunately I have to let you know that your modifications are not working as expected and it still goes from queuing up the DISCONNECT packet to destroying the tlsio connection.  However, I think I have found the problem.

In `DisconnectFromClient()` there is a call to `mqtt_client_disconnect()` which calls the `tlsio_send_async()` interface of the tlsio adapter.  After that it sets `mqtt_client->xioHandle` to `NULL.`  As a consequence of this, all further calls to `mqtt_client_dowork()` have no effect.
    `if (mqtt_client != NULL && mqtt_client->xioHandle != NULL)`

Incidentally, the current behavior of the mqtt transport/mqtt client has revealed another weakness (bug?)  If the connection is not closed (or fails to close), destroying the tlsio connection does not set `mqtt_client->socketConnected` and `mqtt_client->clientConnected `to false.  This can be observed when the SAS token expires.  The tlsio interface is destroyed, then a new one is created and opened.  However, when the `onOpenComplete()` callback is called (with` open_result` = `IO_OPEN_OK`), since `mqtt_client->socketConnected` is still `true`, the connection is immediately closed.  The next time around it works, because closing the connection has set `mqtt_client->socketConnected` back to `false`, but it is a waste of time having to open the connection twice.  I wasn’t sure if I should raise this as a separate issue because it is sort of entangled with the current one.
